packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -0,0 +1,337 @@
+/**
+ * @license Copyright (c) 2003-2021, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module source-editing/sourceediting
+ */
+
+/* global console */
+
+import { Plugin, PendingActions } from 'ckeditor5/src/core';
+import { ButtonView } from 'ckeditor5/src/ui';
+import { createElement, ElementReplacer } from 'ckeditor5/src/utils';
+
+import '../theme/sourceediting.css';
+
+import sourceEditingIcon from '../theme/icons/source-editing.svg';
+
+const COMMAND_FORCE_DISABLE_ID = 'SourceEditingMode';
+
+/**
+ * The source editing feature.
+ *
+ * It provides the possibility to view and edit the source of the document.
+ *
+ * For a detailed overview, check the {@glink features/source-editing source editing feature documentation} and the
+ * {@glink api/source-editing package page}.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class SourceEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'SourceEditing';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ PendingActions ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Flag indicating whether the document source mode is active.
+		 *
+		 * @observable
+		 * @member {Boolean}
+		 */
+		this.set( 'isSourceEditingMode', false );
+
+		/**
+		 * The element replacer instance used to replace the editing roots with the wrapper elements containing the document source.
+		 *
+		 * @private
+		 * @member {module:utils/elementreplacer~ElementReplacer}
+		 */
+		this._elementReplacer = new ElementReplacer();
+
+		/**
+		 * Maps all root names to wrapper elements containing the document source.
+		 *
+		 * @private
+		 * @member {Map.<String,HTMLElement>}
+		 */
+		this._replacedRoots = new Map();
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const t = editor.t;
+
+		editor.ui.componentFactory.add( 'sourceEditing', locale => {
+			const buttonView = new ButtonView( locale );
+
+			buttonView.set( {
+				label: t( 'Source' ),
```

Comment:
```suggestion
				label: t( 'View source' ),
```
WDYT? I am afraid that the `Source` is not explicit enough.

Reply:
I would leave it as it is now (with the "Source" label), because:
* this label was proposed by our UI/UX team,
* we often lack space in the toolbar so IMHO a shorter label is a better label,
* we will be consistent with the source area plugin from CKE4, where there is also "Source" label: https://github.com/ckeditor/ckeditor4/blob/major/plugins/sourcearea/lang/en.js